### PR TITLE
Llama4 Follow-Up PR

### PR DESCRIPTION
**Pull request by jrplatin**
# Description

In this PR, I do the following:
* Rename `MoeBlock` to `RoutedMoE` for clarity
* Rename `DeepSeekMoeBlock` to `RoutedAndSharedMoE` for clarity (especially for Llama4)
* Set `name` values fore existing instances of the above two blocks to ensure backwards compatibility with old checkpoints that use the old naming scheme
* Integrate attention temperature tuning with both Llama4 Scout and Maverick (only NoPE layers use this)
* Integrate chunked attention with both Llama4 Scout and Maverick (NoPE layers use this, other layers use global attention)
* Add `sparse_matmul` and `megablox` support for Llama4

# Tests

**Llama4**
* Re-ran unscanned conversion for the Scout model and ensured matching logits (**NOTE:** this PR will break existing Llama4 checkpoints, but this shouldn't be a major deal for now given the novelty of support)

**DeepSeek**
* Ran forward pass logit checker on an existing DeepSeek-16B (V2) checkpoint

**Mixtral**
* Ran forward pass logit checker on an existing Mixtral 8 x 7B checkpoint 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/configs/base.yml [hunk 1]
```diff
@@ -686,6 +686,13 @@ use_qk_norm: False
 nope_layer_interval: -1
 # Every `X` layers is MoE layer
 interleave_moe_layer_step: 1
+# dynamically scale the attention temperature for each query token based on sequence length
+# Recommended for long sequences (e.g., >32k tokens) to maintain stable output results
+# See (https://arxiv.org/abs/2501.19399) for more details
+temperature_tuning: False
+temperature_tuning_scale: 0.1
+temperature_tuning_floor_scale: 8192.0
```
Comment (gagika) [inline MaxText/configs/base.yml:694]: do you want to do those from config?

if it's not going to change between Llama 4 models, you could pass it in llama4.py file.

If user changes those config it will impact only Llama4 models (perhaps you can comment it if you decide to keep, e.g. Llama4 attention temperature ... )

**Review by gagika — APPROVED** [review]
Thanks, Only one small comment.

File: MaxText/configs/base.yml (hunk 1 above)
Comment (jrplatin) [reply MaxText/configs/base.yml:694]: Good point -- I don't expect these to change, so probably best just to hard-code them

File: MaxText/layers/attentions.py [hunk 2]
```diff
@@ -365,7 +365,7 @@ def generate_attention_mask(
       all_ones = jnp.ones_like(output_mask)
       sliding_mask = jnp.triu(all_ones, -1 * self.sliding_window_size + 1) * jnp.tril(all_ones, self.sliding_window_size - 1)
       output_mask = sliding_mask * output_mask
-    elif self.attention_type == AttentionType.CHUNK and output_mask is not None:
+    elif self.attention_type == AttentionType.CHUNK and output_mask is not None and output_mask.shape[-1] > self.chunk_attn_window_size:
```
Comment (jrplatin) [inline MaxText/layers/attentions.py:368]: @gagika  I added this check to mimic what's [done](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama4/modeling_llama4.py#L791) for Llama4, but let me know if it looks off!

Comment (gagika) [reply MaxText/layers/attentions.py:368]: The intended implementation of chunk attention mask should have matched with causal mask for short sequences,  but let's keep it, I will debug it in a few days.

Comment (jrplatin) [reply MaxText/layers/attentions.py:368]: Sounds good!

Comment (richjames0) [inline MaxText/layers/attentions.py:368]: Tiny nit but this line is a bit rough. Could be mitigated by hoisting the last expression to a variable and could import AttentionType

File: MaxText/layers/deepseek.py [hunk 3]
```diff
@@ -211,7 +211,8 @@ def __call__(
         inputs, self.config, self.mesh, self.quant, decoder_segment_ids, decoder_positions, deterministic, model_mode
     )
 
-    mlp_lnx = moe.DeepSeekMoeBlock(
+    mlp_lnx = moe.RoutedAndSharedMoE(
+        name="DeepSeekMoeBlock_0",
```
Comment (richjames0) [inline MaxText/layers/deepseek.py:215]: Maybe a comment explaining why the mismatch

File: MaxText/layers/llama4.py [hunk 4]
```diff
@@ -177,6 +178,10 @@ def __call__(
         is_nope_layer=self.is_nope_layer,
         use_qk_norm=cfg.use_qk_norm,
         query_pre_attn_scalar=query_pre_attn_scalar,
+        temperature_tuning=cfg.temperature_tuning,
+        temperature_tuning_scale=0.1,
```
Comment (richjames0) [inline MaxText/layers/llama4.py:182]: I don't love seeing values set here. Should they be set in base.yml or similar and we enforce override (no default values)?

File: MaxText/layers/llama4.py [hunk 5]
```diff
@@ -177,6 +178,10 @@ def __call__(
...
+        temperature_tuning_floor_scale=8192.0,
+        attention_type=AttentionType.GLOBAL if self.is_nope_layer else AttentionType.CHUNK, # note: chunk_attn_window_size is set in the config
```
Comment (richjames0) [inline MaxText/layers/llama4.py:184]: I like long lines as much as the next person but maybe put the comment on the line above here

File: MaxText/layers/moe.py [hunk 6]
```diff
@@ -142,7 +142,7 @@ def __call__(self, inputs: Array) -> Tuple[Array, Optional[Array]]:
     return output, pre_bias_logits
 
 
-class MoeBlock(nn.Module):
+class RoutedMoE(nn.Module):
   """Mixture of Experts (MoE) block.
```
Comment (richjames0) [inline MaxText/layers/moe.py:146]: Nit: update description

File: MaxText/layers/moe.py [hunk 7]
```diff
@@ -314,9 +314,17 @@ def permute(self, inputs, gate_logits, pre_bias_logits):
...
     inputs_shape = inputs.shape
-    inputs_2d = jnp.reshape(inputs, (inputs_shape[0] * inputs_shape[1], inputs_shape[2]))
+    bsz_times_seq_len = inputs_shape[0] * inputs_shape[1]
+    inputs_2d = jnp.reshape(inputs, (bsz_times_seq_len, inputs_shape[2]))
     weights, selected_experts = self.get_topk(gate_logits, pre_bias_logits)
 
+    if self.config.decoder_block == "llama4":
+      # weights will be of shape (batch_size, seq_len, num_experts_per_tok)
+      router_scores = jax.nn.sigmoid(weights.astype(jnp.float32))  # weights are top_k_weights here
+      assert router_scores.shape[-1] == 1, "Only top-1 routing is supported for Llama4 for now!"
```
Comment (richjames0) [inline MaxText/layers/moe.py:324]: Nit: assert is meant to be to catch logic/coding errors and not validate parameters. This would ideally be a conditional raise

File: MaxText/layers/moe.py [hunk 8]
```diff
@@ -336,6 +344,9 @@ def unpermute(self, intermediate, sorted_selected_experts, weights, batch_size,
     )
     with jax.named_scope("weight_sum"):
       matmul_precision = lax.Precision(self.config.matmul_precision)
+      if self.config.decoder_block == "llama4":
```
Comment (richjames0) [inline MaxText/layers/moe.py:347]: This isn't your fault as it's done all over the place but it'd be lovely if this were turned by pyconfig into an enum. Feel free to implement!

File: MaxText/layers/moe.py [hunk 9]
```diff
@@ -946,7 +957,7 @@ def __call__(self, inputs):
       return self.dense_matmul(inputs, gate_logits, pre_bias_logits, w0_kernel, w1_kernel, wo_kernel)
 
 
-class DeepSeekMoeBlock(nn.Module):
+class RoutedAndSharedMoE(nn.Module):
   """DeepSeek MoE block, combining shared and routed experts.
```
Comment (richjames0) [inline MaxText/layers/moe.py:961]: Description is not out of date

**Review by richjames0 — COMMENTED** [review]
This is great @jrplatin - thank you! I've put in a few nits and one bigger picture thing is I'd love to see tests added and/or modified in this PR :)

File: MaxText/layers/deepseek.py (hunk 3 above)
Comment (jrplatin) [reply MaxText/layers/deepseek.py:215]: Good point!

File: MaxText/layers/llama4.py (hunk 5 above)
Comment (jrplatin) [reply MaxText/layers/llama4.py:184]: Gotcha

File: MaxText/layers/moe.py (hunk 8 above)
Comment (jrplatin) [reply MaxText/layers/moe.py:347]: Sure, will try my hand at it!

File: MaxText/layers/moe.py (hunk 7 above)
Comment (jrplatin) [reply MaxText/layers/moe.py:324]: Good catch -- I will actually just move this check to `pyconfig.py` so we can fail earlier

File: MaxText/layers/llama4.py (hunk 4 above)
Comment (jrplatin) [reply MaxText/layers/llama4.py:182]: Discussed offline -- will keep for now

**Review by richjames0 — APPROVED** [review]
lgtm

**Review by RissyRan — APPROVED** [review]
LGTM!
